org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/LoggingMessageConsumerWrapper.java

```diff
@@ -0,0 +1,132 @@
+        return new MessageConsumer() {
+            @Override
+            public void consume(Message message) throws MessageIssueException, JsonRpcException {
+                final String action;
+                final String method;
+                final String traceLine;
+                final String logLine;
+                final long now = System.currentTimeMillis();
+                final String date = dateFormat.format(new Date(now));
+
+                if (messageConsumer instanceof StreamMessageConsumer) {
+                    if (message instanceof RequestMessage) {
+                        RequestMessage requestMessage = (RequestMessage) message;
+                        action = "Sending request";
+                        method = requestMessage.getMethod();
+                        String id = requestMessage.getId();
+                        idsToMethods.put(id, method);
+                        idsToRequestStartTimes.put(id, now);
+                        traceLine = String.format("[Trace - %s] %s '%s - (%s)'", date, action, method, id);
+
+                        Object params = requestMessage.getParams();
+                        String paramsJson = MessageJsonHandler.toString(params);
+                        logLine = String.format("Params: %s", paramsJson);
+                    } else if (message instanceof ResponseMessage) {
+                        ResponseMessage responseMessage = (ResponseMessage) message;
+                        action = "Sending response";
+                        String id = responseMessage.getId();
+                        method = idsToMethods.get(id);
+
+                        long requestStartMillis = idsToRequestStartTimes.get(id);
+                        long latencyMillis = now - requestStartMillis;
+                        String latency = String.format("%sms", latencyMillis);
+                        traceLine = String.format("[Trace - %s] %s '%s - (%s)' took %s", date, action, method, id, latency);
+
+                        Object result = responseMessage.getResult();
+                        String resultJson = MessageJsonHandler.toString(result);
+                        logLine = String.format("Result: %s", resultJson);
+                    } else if (message instanceof NotificationMessage) {
+                        NotificationMessage notificationMessage = (NotificationMessage) message;
+                        action = "Sending notification";
+                        method = notificationMessage.getMethod();
+                        traceLine = String.format("[Trace - %s] %s '%s'", date, action, method);
+
+                        Object params = notificationMessage.getParams();
+                        String paramsJson = MessageJsonHandler.toString(params);
+                        logLine = String.format("Params: %s", paramsJson);
+                    } else {
+                        throw new IllegalStateException("Unhandled message: " + message);
```

Comment:
I would rather print an error to the output `printWriter` than throw an exception here. Same for other throws below.